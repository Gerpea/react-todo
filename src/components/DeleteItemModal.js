import React from 'react'
import PropTypes from 'prop-types'

import { Modal, Button } from 'react-bootstrap'

const DeleteItemModal = ({ todo, onDelete, onHide, ...rest }) => {
  return (
    <Modal {...rest} animation={false} onHide={onHide}>
      <Modal.Header>
        <Modal.Title>Вы уверены, что хотите удалить элемент?</Modal.Title>
      </Modal.Header>
      <Modal.Footer>
        <Button
          variant='primary'
          onClick={() => {
            onHide?.call(undefined)
            onDelete?.call(undefined, todo.uid)
          }}
          data-testid='delete-todo'>
          Да
        </Button>
        <Button
          variant='secondary'
          onClick={() => onHide?.call(undefined)}
          data-testid='not-delete-todo'>
          Нет
        </Button>
      </Modal.Footer>
    </Modal>
  )
}

DeleteItemModal.propTypes = {
  todo: PropTypes.object,
  onDelete: PropTypes.func,
  onHide: PropTypes.func,
}

DeleteItemModal.defaultProps = {
  todo: {},
  onDelete: () => {},
  onHide: () => {},
}

export default DeleteItemModal
