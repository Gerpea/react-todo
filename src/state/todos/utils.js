function removeTodo(state, uid) {
  const children = getChildren(state, uid)

  let withoutChild = state
  children.forEach((child) => {
    withoutChild = removeTodo(state, child.uid)
  })

  return withoutChild.reduce((acc, cur) => {
    if (cur.uid !== uid) {
      acc.push(cur)
    }
    return acc
  }, [])
}

function getChildren(state, uid) {
  return state.filter((item) => item.parentUid === uid)
}

function getParentCount(state, uid) {
  const parentUid = state.find((item) => item.uid === uid)?.parentUid
  if (parentUid) {
    return 1 + getParentCount(state, parentUid)
  } else {
    return 0
  }
}

export { removeTodo, getChildren, getParentCount }
