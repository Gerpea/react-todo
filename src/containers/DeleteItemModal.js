import { connect } from 'react-redux'

import { removeTodo } from '../state/todos/actions'

import DeleteItemModal from '../components/DeleteItemModal'

const mapDispatchToProps = {
  onDelete: (uid) => removeTodo(uid),
}

export default connect(null, mapDispatchToProps)(DeleteItemModal)
