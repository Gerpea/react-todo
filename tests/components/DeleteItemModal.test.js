import React from 'react'
import ReactDOM from 'react-dom'
import { emptyWrapper, render, cleanup, fireEvent } from '../utils'
import renderer from 'react-test-renderer'

import DeleteItemModal from '../../src/components/DeleteItemModal'

afterEach(cleanup)

it('renders without crashing', () => {
  render(<DeleteItemModal />)
})

ReactDOM.createPortal = (node) => node
it('matches snapshot', () => {
  const container = document.createElement('div')
  const tree = renderer
    .create(ReactDOM.createPortal(<DeleteItemModal show={true} />, container))
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('should call onHide when cancel button clicked', () => {
  const onHide = jest.fn()
  const leftButton = { button: 0 }
  const { getByTestId } = render(<DeleteItemModal show={true} onHide={onHide} />)

  fireEvent.click(getByTestId('not-delete-todo'), leftButton)

  expect(onHide.mock.calls.length).toBe(1)
})

describe('onDelete', () => {
  it('should be called when delete button clicked', () => {
    const onDelete = jest.fn()
    const leftButton = { button: 0 }
    const { getByTestId } = render(
      emptyWrapper(<DeleteItemModal show={true} onDelete={onDelete} />)
    )

    fireEvent.click(getByTestId('delete-todo'), leftButton)

    expect(onDelete.mock.calls.length).toBe(1)
  })
})
