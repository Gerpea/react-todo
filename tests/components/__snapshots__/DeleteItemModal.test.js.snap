// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <div
    className="modal-backdrop show"
    onClick={[Function]}
  />,
  <div
    aria-modal={true}
    className="modal"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <div
            className="modal-title h4"
          >
            Вы уверены, что хотите удалить элемент?
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-primary"
            data-testid="delete-todo"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Да
          </button>
          <button
            className="btn btn-secondary"
            data-testid="not-delete-todo"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Нет
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
